R/derive_vars_dt_dtm_utils.R

```diff
@@ -260,3 +262,515 @@ get_partialdatetime <- function(dtc) {
+  if (is_datetime) dtm_level(highest_imputation) else dt_level(highest_imputation)
+}
+
+#' Get Imputation Targets
+#'
+#' @description
+#' `r lifecycle::badge("stable")`
+#'
+#' This is a helping function for `get_dt_dtm_range()`
+#' Determines the imputation targets for date and time components.
+#'
+#' @param partial A list of partial date/time components.
+#' @param date_imputation A character specifying the date imputation method.
+#' @param time_imputation A character specifying the time imputation method.
+#' @param is_datetime A logical indicating whether it's a datetime imputation.
+#'
+#' @returns A list of imputation targets for date and (if applicable) time components.
+#'
+#' @keywords internal
+get_imputation_targets <- function(partial, date_imputation, time_imputation, is_datetime) {
+  target_date <- get_imputation_target_date(
+    date_imputation = date_imputation,
+    month = partial[["month"]]
+  )
+
+  if (is_datetime) {
+    target_time <- get_imputation_target_time(time_imputation = time_imputation)
+    return(c(target_date, target_time))
+  }
+
+  return(target_date)
+}
+
+#' Adjust Last Day Imputation
+#'
+#' @description
+#' `r lifecycle::badge("stable")`
+#'
+#' This is a helping function for `get_dt_dtm_range()`
+#' Adjusts the imputed date/datetime to the last day of the month when necessary.
+#'
+#' @param imputed_dtc A character vector of imputed date/datetime strings.
+#' @param partial A list of partial date/time components.
+#' @param is_datetime A logical indicating whether it's a datetime adjustment.
+#'
+#' @returns A character vector of adjusted date/datetime strings.
+#'
+#' @details
+#' This function is used when the date imputation method is set to "last" and
+#' the day component is missing. It adjusts the date to the last day of the month.
+#'
+#' @keywords internal
+#'
+#' @importFrom lubridate ymd_hms ymd rollback
+adjust_last_day_imputation <- function(imputed_dtc, partial, is_datetime) {
+  if_else(
+    is.na(partial[["day"]]),
+    if (is_datetime) {
+      strftime(
+        rollback(ymd_hms(imputed_dtc) + months(1)),
+        format = "%Y-%m-%dT%H:%M:%S",
+        tz = "UTC"
+      )
+    } else {
+      strftime(
+        rollback(ymd(imputed_dtc) + months(1)),
+        format = "%Y-%m-%d"
+      )
+    },
+    imputed_dtc
+  )
+}
+
+
+#' Impute Missing Values
+#'
+#' @description
+#' `r lifecycle::badge("stable")`
+#'
+#' This is a helping function for `get_dt_dtm_range()`
+#' Imputes missing values in partial date/time components using target values.
+#'
+#' @param partial A list of partial date/time components.
+#' @param target A list of target values for imputation.
+#' @param components A character vector of component names.
+#'
+#' @returns A list of imputed date/time components.
+#'
+#' @keywords internal
+impute_values <- function(partial, target, components) {
+  imputed <- vector("list", length(components))
```

Comment:
Maybe we should check that `partial` and `target` are consistent, i.e., that they have the same names.